File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task138.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

/**
 * Manages user roles and permissions securely.
 * - Encapsulates data within private members.
 * - Uses const references for safe and efficient parameter passing.
 * - Validates input by checking for empty strings.
 * - Uses unordered_map and unordered_set for efficient lookups.
 */
class PermissionManager {
private:
    // Maps role names to a set of permission strings
    std::unordered_map<std::string, std::unordered_set<std::string>> rolesToPermissions;
    // Maps user names to a set of role strings
    std::unordered_map<std::string, std::unordered_set<std::string>> usersToRoles;

public:
    PermissionManager() = default;

    /**
     * Adds a new role to the system.
     * @param roleName The name of the role to add.
     * @return true if successful, false if input is invalid or role exists.
     */
    bool addRole(const std::string& roleName) {
        if (roleName.empty()) return false;
        // The insert method returns a pair; the .second is a bool indicating if insertion took place.
        return rolesToPermissions.insert({roleName, {}}).second;
    }

    /**
     * Adds a new user to the system.
     * @param userName The name of the user to add.
     * @return true if successful, false if input is invalid or user exists.
     */
    bool addUser(const std::string& userName) {
        if (userName.empty()) return false;
        return usersToRoles.insert({userName, {}}).second;
    }

    /**
     * Adds a permission to an existing role.
     * @param roleName The role to modify.
     * @param permission The permission to add.
     * @return true if successful, false if role doesn't exist or input is invalid.
     */
    bool addPermissionToRole(const std::string& roleName, const std::string& permission) {
        if (roleName.empty() || permission.empty()) return false;
        auto it = rolesToPermissions.find(roleName);
        if (it == rolesToPermissions.end()) {
            return false; // Role not found
        }
        it->second.insert(permission);
        return true;
    }

    /**
     * Assigns an existing role to an existing user.
     * @param userName The user to whom the role will be assigned.
     * @param roleName The role to assign.
     * @return true if successful, false if user or role doesn't exist or input is invalid.
     */
    bool assignRoleToUser(const std::string& userName, const std::string& roleName) {
        if (userName.empty() || roleName.empty()) return false;
        auto userIt = usersToRoles.find(userName);
        if (userIt == usersToRoles.end() || rolesToPermissions.find(roleName) == rolesToPermissions.end()) {
            return false; // User or role not found
        }
        userIt->second.insert(roleName);
        return true;
    }

    /**
     * Checks if a user has a specific permission.
     * @param userName The user to check.
     * @param permission The permission to check for.
     * @return true if the user has the permission through any of their roles, false otherwise.
     */
    bool checkPermission(const std::string& userName, const std::string& permission) {
        if (userName.empty() || permission.empty()) return false;

        auto userIt = usersToRoles.find(userName);
        if (userIt == usersToRoles.end()) {
            return false; // User not found
        }

        for (const auto& role : userIt->second) {
            auto roleIt = rolesToPermissions.find(role);
            if (roleIt != rolesToPermissions.end()) {
                if (roleIt->second.count(permission)) {
                    return true;
                }
            }
        }
        return false;
    }
};

int main() {
    PermissionManager manager;

    // Setup
    manager.addRole("admin");
    manager.addRole("editor");
    manager.addRole("viewer");

    manager.addPermissionToRole("admin", "read");
    manager.addPermissionToRole("admin", "write");
    manager.addPermissionToRole("admin", "delete");
    manager.addPermissionToRole("admin", "manage_users");

    manager.addPermissionToRole("editor", "read");
    manager.addPermissionToRole("editor", "write");

    manager.addPermissionToRole("viewer", "read");

    manager.addUser("alice");
    manager.addUser("bob");
    manager.addUser("charlie");

    manager.assignRoleToUser("alice", "admin");
    manager.assignRoleToUser("bob", "editor");
    manager.assignRoleToUser("charlie", "viewer");

    std::cout << "--- Running Test Cases ---" << std::endl;
    std::cout << std::boolalpha; // Print bools as "true" or "false"

    // Test Case 1: Admin has delete permission
    std::cout << "1. Alice has 'delete' permission: " << manager.checkPermission("alice", "delete") << std::endl;

    // Test Case 2: Editor has write permission
    std::cout << "2. Bob has 'write' permission: " << manager.checkPermission("bob", "write") << std::endl;

    // Test Case 3: Editor does NOT have delete permission
    std::cout << "3. Bob has 'delete' permission: " << manager.checkPermission("bob", "delete") << std::endl;

    // Test Case 4: Non-existent user
    std::cout << "4. Dave has 'read' permission: " << manager.checkPermission("dave", "read") << std::endl;
    
    // Test Case 5: User with multiple roles
    manager.assignRoleToUser("bob", "viewer"); // bob is now editor and viewer
    std::cout << "5. Bob (now editor & viewer) has 'read' permission: " << manager.checkPermission("bob", "read") << std::endl;

    std::cout << "--- Test Cases Finished ---" << std::endl;

    return 0;
}

File: dataset/CoT/GEMINI_gemini-2.5-pro/Task138.py
class PermissionManager:
    """Manages user permissions and privileges."""

    def __init__(self):
        # Using dictionaries to store roles and user assignments
        # roles: {"role_name": {"permission1", "permission2"}}
        # users: {"user_name": "role_name"}
        self._roles = {}
        self._users = {}

    def add_role(self, role_name, permissions):
        """Adds a new role with a set of permissions."""
        if isinstance(role_name, str) and role_name and isinstance(permissions, set):
            self._roles[role_name] = permissions.copy()

    def assign_role_to_user(self, user_name, role_name):
        """Assigns an existing role to a user."""
        if isinstance(user_name, str) and user_name and role_name in self._roles:
            self._users[user_name] = role_name

    def check_permission(self, user_name, permission):
        """
        Checks if a user has a specific permission.

        Args:
            user_name (str): The name of the user.
            permission (str): The permission to check.

        Returns:
            bool: True if the user has the permission, False otherwise.
        """
        if not isinstance(user_name, str) or not isinstance(permission, str):
            return False

        # Safely get the user's role
        role_name = self._users.get(user_name)
        if not role_name:
            return False

        # Safely get the permissions for that role
        role_permissions = self._roles.get(role_name)
        if not role_permissions:
            return False

        # Check if the permission is in the role's permission set
        return permission in role_permissions

def main():
    """Main function to run test cases."""
    manager = PermissionManager()

    # 1. Define permissions and roles
    permissions = {
        "ADMIN": {"READ", "WRITE", "DELETE", "EXECUTE"},
        "EDITOR": {"READ", "WRITE"},
        "VIEWER": {"READ"}
    }
    
    for role, perms in permissions.items():
        manager.add_role(role, perms)

    # 2. Create users and assign roles
    manager.assign_role_to_user("alice", "ADMIN")
    manager.assign_role_to_user("bob", "EDITOR")
    manager.assign_role_to_user("charlie", "VIEWER")

    print("Running test cases...\n")

    # Test Case 1: Admin user checking an admin-only permission
    print("Test 1: Does alice (ADMIN) have 'DELETE' permission?")
    print(f"Result: {manager.check_permission('alice', 'DELETE')}") # Expected: True

    # Test Case 2: Editor user checking an editor permission
    print("\nTest 2: Does bob (EDITOR) have 'WRITE' permission?")
    print(f"Result: {manager.check_permission('bob', 'WRITE')}") # Expected: True

    # Test Case 3: Editor user checking an admin-only permission
    print("\nTest 3: Does bob (EDITOR) have 'DELETE' permission?")
    print(f"Result: {manager.check_permission('bob', 'DELETE')}") # Expected: False

    # Test Case 4: Viewer user checking a read-only permission
    print("\nTest 4: Does charlie (VIEWER) have 'READ' permission?")
    print(f"Result: {manager.check_permission('charlie', 'READ')}") # Expected: True

    # Test Case 5: Non-existent user checking any permission
    print("\nTest 5: Does david (non-existent) have 'READ' permission?")
    print(f"Result: {manager.check_permission('david', 'READ')}") # Expected: False

if __name__ == "__main__":
    main()